More object-oriented python jobs

paulfantom opened the pull request:
Effect of talks from #555.

Major changes:
- moved configuration schema to YAML so `pyyaml` is needed.
- modules present only definition of a service, `python.d.plugin` reads YAML configuration and runs as many jobs as needed.
- every Service class must be a child of BaseService. BaseService handles `update_every`, `retries`, `priority`, and dynamically created `timetable`.
- rewritten `mysql.chart.py` and `example.chart.py` to support those changes.

Minor changes:
- better handling of disabled modules
- probably some more, but I don't remember now

I have tested some cases, but probably not enough.

ktsaou commented:
Nice work!

I am reviewing it now...

paulfantom commented:
Just found an issue with pyyaml. It has to be installed system-wide.
Python can import modules when they are in directories specified in `$PYTHONPATH` environment variable. I am appending `python_modules` directory in line 407 (previously 348) of `python.d.plugin`, but pyyaml is imported at the beginning. If I move extending of `$PYTHONPATH` before `import pyyaml`, it will break ability to set `plugins_dir`.

ktsaou commented:
ok.

To install pyaml for python3 on fedora24 I had to do this:

```sh
curl https://bootstrap.pypa.io/get-pip.py | sudo python3
```

This provided `pip3`. Then I get this:

```sh
$ ./plugins.d/python.d.plugin debug 1
Traceback (most recent call last):
  File "./plugins.d/python.d.plugin", line 416, in <module>
    run()
  File "./plugins.d/python.d.plugin", line 370, in run
    modules_conf = conf['plugins_config_dir']
TypeError: string indices must be integers
```

paulfantom commented:
It is not pyaml it is pyyaml.
double 'y'.

paulfantom commented:
And it should be in main fedora repository.

ktsaou commented:
My bad. Yes it is `pyyaml`. I uninstalled `pyaml`, `pyyaml` and re-installed `pyyaml`. Same error.

ktsaou commented:
Without pyyaml installed I get this:
```sh
$ ./plugins.d/python.d.plugin debug 1
python.d.plugin: Cannot find yaml library
DISABLE
```

So, pyyaml is there.

paulfantom commented:
What is in your `python.d.conf`?

ktsaou commented:
```sh
# pip3 install pyyaml
Collecting pyyaml
Installing collected packages: pyyaml
Successfully installed pyyaml-3.11
```

Is 3.11 good?

paulfantom commented:
Yes, this version is ok.
I am trying to reproduce your error on ubuntu 16.04. And I can't. 
Could you add `print(conf)` after `conf = read_config(configfile)` (line 362) and run?

ktsaou commented:
```sh
$ ./plugins.d/python.d.plugin debug 1
example=no
Traceback (most recent call last):
  File "./plugins.d/python.d.plugin", line 417, in <module>
    run()
  File "./plugins.d/python.d.plugin", line 371, in run
    modules_conf = conf['plugins_config_dir']
TypeError: string indices must be integers
```

ktsaou commented:
Oh... it is pyaml parsing error. I did not install it.

ktsaou commented:
Yes. After installing it and replacing `/etc/netdata/python.d.conf` is worked.

paulfantom commented:
Oh. I should add some information about it.

ktsaou commented:
The same when loading modules config from the previous version:

```sh
$ ./plugins.d/python.d.plugin debug 1
{'mysql': True, 'example': True}
python.d: started from ./plugins.d/python.d.plugin with options: debug 1
python.d: loading chart: '/home/costa/src/netdata-paulfantom.git/python.d/example.chart.py'
mysql.chart.py: You need to install MySQLdb or PyMySQL module to use mysql.chart.py plugin
python.d: 
python.d: loading chart options: '/etc/netdata/python.d/example.conf'
python.d: something went wrong while loading configuration 'str' object has no attribute 'pop'
Traceback (most recent call last):
  File "./plugins.d/python.d.plugin", line 417, in <module>
    run()
  File "./plugins.d/python.d.plugin", line 411, in run
    charts = PythonCharts(interval, modules, modules_dir, modules_conf, disabled)
  File "./plugins.d/python.d.plugin", line 45, in __init__
    self.jobs = self._create_jobs(configured_modules)
  File "./plugins.d/python.d.plugin", line 59, in _create_jobs
    for name in module.config:
AttributeError: module 'example' has no attribute 'config'
```

paulfantom commented:
It is the same loader function.

ktsaou commented:
example plugin works.
I am trying to install mysqldb under python3. Its pip name is `mysqlclient`?

paulfantom commented:
I think it is `MySQL-python`

ktsaou commented:
Well, I cannot get it:

```sh
# pip3 install mysqlclient
Collecting mysqlclient
  Using cached mysqlclient-1.3.7.tar.gz
    Complete output from command python setup.py egg_info:
    /bin/sh: mysql_config: command not found
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-bewwom8q/mysqlclient/setup.py", line 17, in <module>
        metadata, options = get_config()
      File "/tmp/pip-build-bewwom8q/mysqlclient/setup_posix.py", line 44, in get_config
        libs = mysql_config("libs_r")
      File "/tmp/pip-build-bewwom8q/mysqlclient/setup_posix.py", line 26, in mysql_config
        raise EnvironmentError("%s not found" % (mysql_config.path,))
    OSError: mysql_config not found
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-bewwom8q/mysqlclient/
```

```sh
# pip3 install mysql-python
Collecting mysql-python
  Downloading MySQL-python-1.2.5.zip (108kB)
    100% |████████████████████████████████| 112kB 261kB/s 
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-build-qv4b89__/mysql-python/setup.py", line 13, in <module>
        from setup_posix import get_config
      File "/tmp/pip-build-qv4b89__/mysql-python/setup_posix.py", line 2, in <module>
        from ConfigParser import SafeConfigParser
    ImportError: No module named 'ConfigParser'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-qv4b89__/mysql-python/
```

ktsaou commented:
pymysql works.

ktsaou commented:
When `python.d.plugin` is run with a module name that does not exist, it should exit with `DISABLE`. Now it just ignores this parameter and continues.
We might need this behavior in the future to allow multiple python.d.plugins running in different namespaces / containers.

ktsaou commented:
I opened python.d.plugin in Clion and it gives 61 errors and 4 warnings. I don't know how to give them to you though. Can you install the evaluation version of Clion and check the errors and warnings? Probably they are nothing, but I thought I should let you know.

ktsaou commented:
An example:

![screenshot from 2016-06-19 14-43-47](https://cloud.githubusercontent.com/assets/2662304/16177097/5bd9b980-362c-11e6-876e-fce22473cd87.png)

paulfantom commented:
Probably artifacts from previous version. I will fix them.
Also `mysqldb` in pypi is not supported for python3 (https://pypi.python.org/pypi/MySQL-python/1.2.4)
There should be a good version in system repository (on ubuntu it is named: python3-mysqldb), and on github: (https://github.com/PyMySQL/mysqlclient-python)

I will add 'DISABLE' on `run()` exit.

ktsaou commented:
I see that JetBrains has also PyCharm IDE for python. JetBrains has donated 4 licenses to firehol.org, so I can give you one license for the time you work with us. Do you want it? I'll need your email - email it to me to costa@tsaousis.gr

ktsaou commented:
I'll be away for a few hours.
Let me know when you are ready.

paulfantom commented:
Ok, ready to be tested. 
I have also added some experimental python2 support

paulfantom commented:
Tested:
- `example.chart.py`
- `mysql.chart.py`
- malformed configuration
- recovering from `update()` failure [ retries ]
- running without modules
- `python2 python.d.plugin debug example`
- probably some more minor tests

Since I don't know what more to test, it is your turn now.

Also what should I do with pyyaml issue mentioned at the beginning?

Fun fact: every error displayed in Clion, PyCharm shows as a warning.
And I got rid of as many warnings as it was possible.

ktsaou commented:
Is it now python2 compatible? It seems running...

If it is, I suggest to change the shebang to `python` and remove the relative info from `python.d/README.md`.

ktsaou commented:
> Also what should I do with pyyaml issue mentioned at the beginning?

What I have understood is that pyyaml must be installed system-wide to be used. This might be a problem when we start releasing binary packages. We will have to have a dependency to that which might not be possible on all distros.

Is there anything we can do to allow us having in `python.d/python_modules` too?

ktsaou commented:
Please log job names with errors. I have this `/etc/netdata/python.d/mysql.conf`:

```yaml
$ cat /etc/netdata/python.d/mysql.conf 
# Example configuration of mysql.chart.py
# YAML format

update_every: 5

local: 
  'my.cnf' : '/etc/mysql/my.cnf'

local_s:
  user     : 'root'
  password : ''
  socket   : '/var/run/mysqld/mysqld.sock'

remote:
  user     : 'root'
  password : ''
  host     : '127.0.0.1'
  port     : '3306'
```

Check the error messages:

```sh
 ./plugins.d/python.d.plugin debug >/dev/null
python.d.plugin: Using python 3
python.d: started from ./plugins.d/python.d.plugin with options: debug
python.d: loading chart: '/home/costa/src/netdata-paulfantom.git/python.d/example.chart.py'
mysql.chart.py: using MySQLdb
python.d: loading chart: '/home/costa/src/netdata-paulfantom.git/python.d/mysql.chart.py'
python.d: loading chart options: '/etc/netdata/python.d/example.conf'
python.d: loading chart options: '/etc/netdata/python.d/mysql.conf'
mysql.chart.py has problem connecting to server: (2002, 'Can\'t connect to local MySQL server through socket \'\' (111 "Connection refused")')
python.d: chart 'mysql_local' check() function reports failure.
mysql.chart.py has problem connecting to server: (2002, 'Can\'t connect to local MySQL server through socket \'/var/run/mysqld/mysqld.sock\' (2 "No such file or directory")')
python.d: mysql_local_s is misbehaving. Reason: argument of type 'NoneType' is not iterable
```

I got 2 errors (but I was expecting 3, since I have 3 jobs). Each error should state the job name so that I can easily figure out which one is problematic.

paulfantom commented:
Fixed lack of errors.
Job names were always logged, but I changed them to be more visible.
Now it runs on python in version default to platform.

About yaml and packaging.
I am trying to use only packages which are in default repositories on openSUSE, and ubuntu (probably on Fedora too). This way packaging will be very easy, just add a dependency and voila.
Python uses modules which it can find in `PYTHONPATH` environment variable. I am appending absolute path to `python_modules` to this variable just before creating instance of PythonCharts. It cannot be done earlier because it needs to check if there is an alternative path, which is taken from configuration file under variable named `plugins_dir`. But to read configuration file PyYAML is needed, which can be in `python_modules`. And we end with some sort of circular dependency.
Simplest way to remedy this would be to remove `plugins_dir` from configuration file. Then at the beginning of `python.d.plugin` append to `PYTHONPATH` path relative to `NETDATA_PLUGINS_DIR`

ktsaou commented:
Well... having dependencies system-wide is good for most up to date installations. I think however, that we will be forced to provide a `.tar.gz` package that should be unpacked in `/opt/netdata`, having everything pre-installed there. Can we have an option or something that will allow us to provide pyyaml?

paulfantom commented:
I think that it would be better to preserve current directory schema and assume `python_modules` are always in the same place relative to NETDATA_PLUGINS_DIR. This way we can put pyyaml into `python_modules` for installation in /opt or use packages from system repo.

ktsaou commented:
it sounds good.
Are you ready to merge this?

paulfantom commented:
I need to test environment variable handling and pyyaml in `python_modules`.

ktsaou commented:
ok. let me know when you are ready.

paulfantom commented:
I think it is ready. I have added simple script to simplify installing packages with `pip`

ktsaou commented:
merged! Thanks!
